### PR TITLE
Better viewer zoom/center sync with different refdata

### PR DESCRIPTION
**Pull request by bmorris3**
At time of writing, https://github.com/spacetelescope/jdaviz/pull/2179 un-syncs the pan and zoom between viewers in Imviz when different reference data are selected in each viewer, and re-syncs them when the reference data are made to match.

This PR syncs the sky centroids of the FOV in each viewer, and approximately syncs the zoom level. I say the zoom sync is "approximate" because we can't uniformly assign one set of `{x/y}_{min/max}` for all viewers when any are allowed to have different reference data (/rotation). Instead, I compute the average width of each side of the active viewer in world coordinates, and set the zoom level in the other viewers to match the active viewer's width. 

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-3643)

### CODE REVIEW

File: jdaviz/configs/imviz/plugins/tools.py [hunk 1]
```diff
@@ -20,24 +20,7 @@ class _ImvizMatchedZoomMixin(_MatchedZoomMixin):
...
-        # If adding the data to the viewer, make sure it is not actually shown since the
-        # user didn't request it.
-        for viewer in self._iter_matched_viewers(include_self=False):
-            if self.viewer.state.reference_data not in viewer.state.layers_data:
-                viewer.add_data(self.viewer.state.reference_data)
-                for layer in viewer.state.layers:
-                    if layer.layer is self.viewer.state.reference_data:
-                        layer.visible = False
-                        break
-            viewer.state.reference_data = self.viewer.state.reference_data
```
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/tools.py:40]: so glad to finally be rid of this - thanks!

**Comment by pllim** [comment]
As I mentioned at tag-up, since this could be confusing and has confused you more than once, is it less confusing to force the other viewers to switch to the same rotation as the viewer where this button is clicked first before link pan/zoom is executed?

Would anyone sane want to link pan/zoom on data rotated differently in the sky? Usually when I do it, I expect the data to be aligned (by pixel if I am inspecting data quality flags or by sky if I am inspecting an astronomical object of interest).

**Comment by bmorris3** [comment]
@pllim The first example that comes to mind is something like: "I have two galaxies in my image. I want to make elliptical subsets for each, aligning the major axis of the ellipse with the major axis of the galaxy. I open the same image in two viewers with the correct rotation for each galaxy, and draw the elliptical subset in each viewer (so I can avoid changing subset rotation via the plugin)."

**Comment by pllim** [comment]
Re: https://github.com/bmorris3/jdaviz/pull/12#issuecomment-1766677109

Why would linked pan/zoom be useful in that situation?

**Comment by bmorris3** [comment]
@pllim When you load the images, in general, the pan and zoom won't be ideal for drawing your subsets, so you'll probably want to center and zoom into the region with your targets.

**Comment by pllim** [comment]
From "offline" conversations, Brett and I agree that these are viable alternative solutions that would be less confusing but need other devs or POs inputs. Though what I wrote here is a bit more than what was on Slack because my ideas evolved a bit as I typed.

1. Force all other viewers to switch to using reference image of the viewer where the button is activated and used. This might be complicated if we allow the same button to be active on different viewers at the same time?
2. Only execute linked pan/zoom on viewers with matching reference data. We need to know which viewer is under the mouse and whether link pan/zoom is asked for in that viewer. Then we look for only other viewers with the same reference data. 
3. Only execute linked pan/zoom if the reference data is the same across all viewers. That is, this button is disabled altogether unless all viewers have the same reference data.

**Comment by kecnry** [comment]
> Force all other viewers to switch to using reference image of the viewer where the button is activated and used. This might be complicated if we allow the same button to be active on different viewers at the same time?
> Only execute linked pan/zoom on viewers with matching reference data. We need to know which viewer is under the mouse and whether link pan/zoom is asked for in that viewer. Then we look for only other viewers with the same reference data.

I personally think either of these would be more confusing/jarring than a not-exact match of zoom limits.

> Only execute linked pan/zoom if the reference data is the same across all viewers. That is, this button is disabled altogether unless all viewers have the same reference data.

This is an option, but is it worth disabling just for some slight approximation.  Can we just update the tooltip in these cases instead to include some text for "approximate zoom-level matching"?

File: jdaviz/core/tools.py [hunk 2]
```diff
@@ -93,12 +88,32 @@ def deactivate(self):
...
+                # rescale the zoom and center allowing for different
+                # viewer rotations:
+                orig_fov_sky = self.viewer._get_fov(orig_refdata)
+                to_fov_sky = viewer._get_fov(orig_refdata)
+                sky_cen = self.viewer._get_center_skycoord(orig_refdata)
+
+                viewer.zoom(
+                    float(to_fov_sky / orig_fov_sky)
+                )
+                viewer.center_on(sky_cen)
```
Comment (kecnry) [inline jdaviz/core/tools.py:113]: are these processed sequentially?  If they're just modifying the limits under the hood, does wrapping in `delay_callback` work/help?

File: jdaviz/core/tools.py [hunk 3]
```diff
@@ -93,12 +88,32 @@ def deactivate(self):
...
 
         for viewer in self._iter_matched_viewers(include_self=False):
             # orig_lims: limits in this "matched" viewer
             # to_lims: proposed new limits for this "matched" viewer
             orig_lims = {k: getattr(viewer.state, k) for k in self.match_keys}
             to_lims = self._map_limits(self.viewer, viewer, from_lims)
+
+            to_refdata = viewer.state.reference_data
+
+            if orig_refdata != to_refdata and hasattr(self.viewer, '_get_fov'):
```
Comment (kecnry) [inline jdaviz/core/tools.py:102]: this mixin is also used for spectrum-viewers... is this useful there too or do we need to check the viewer type?

**Review by kecnry — APPROVED** [review]
I think this is a good improvement for now and we can always tweak based on user feedback.

https://github.com/bmorris3/jdaviz/assets/877591/c88066a2-f278-4da1-8535-2909612cdc67

File: jdaviz/core/tools.py (hunk 2 above)
Comment (bmorris3) [reply jdaviz/core/tools.py:113]: I didn't expect it to matter or make a difference.

File: jdaviz/core/tools.py (hunk 3 above)
Comment (bmorris3) [reply jdaviz/core/tools.py:102]: The `hasattr` check only passes for image viewers, so this shouldn't be a problem.

**Pull request merged**